### PR TITLE
Add postgresql_tablespace module to be able to create postgres tables…

### LINKED ISSUE #5954: postgresql_db module tablespace parameter
##### Issue Type:

Feature Idea
##### Ansible Version:

ansible 1.4.4
##### Environment:

Running Ansible from Arch Linux

Managing Centos 6/Debian 7
##### Summary:

I would like to make use of this module, but I can't because I can't specify a tablespace for the database.

I noticed this pull request https://github.com/ansible/ansible/pull/4994 but it doesn't seem like the author resubmitted it and also it doesn't seem to modify the postgresql_db module to allow specifying a tablespace anyway.

If I have time i'll try to create a pull request which adds this parameter (should be pretty easy to add).

I just wanted to get this issue in in case someone else has time to do it before me.
##### Steps To Reproduce:

```
tasks:
    - name: create database
      action: postgresql_db name=database1 owner=owner1 tablespace=my_tablespace
```
##### Expected Results:

a new database is created with the specific tablespace my_tablespace
##### Actual Results:

tablespace parameter not possible

### PR DESCRIPTION
**Pull request by antoinell**
SUMMARY
Add support to create postgres tablespaces in ansible.
Fixes #5954

ISSUE TYPE
Feature Pull Request
COMPONENT NAME
postgresql_db

ADDITIONAL INFORMATION
related to #50406 
similar to #23603

### CODE REVIEW

**Comment by Andersson007** [comment]
Hello, @antoinell

**1.What about ALTER TABLESPACE ?** Probably better to add this functionality right now.
```
ALTER TABLESPACE name RENAME TO new_name
ALTER TABLESPACE name OWNER TO new_owner
ALTER TABLESPACE name SET ( tablespace_option = value [, ... ] )
ALTER TABLESPACE name RESET ( tablespace_option [, ... ] )
```

**2. author: "Ansible Core Team"**, Are you sure? I see the "new contributor" status of PR.

**3. Better to add more examples**, for every case

**4. Also it makes sence to add tablespace_option and ownership**(https://www.postgresql.org/docs/11/sql-createtablespace.html)
```
CREATE TABLESPACE tablespace_name
    [ OWNER { new_owner | CURRENT_USER | SESSION_USER } ]
    LOCATION 'directory'
    [ WITH ( tablespace_option = value [, ... ] ) ]
```

**5. Probably better to collect info about the current tablespace and return it by the RETURN block.**
So we could see and use this information by passing the '-vvv' ansible-playbook param.

**6. do_with_password() seems is not used there**

**7. Rename variable to tablespace_info in the tablespace_create() as you did in the next function**
```
112         db_info = get_tablespace_info(cursor, tablespace)
```

**8. It should be changed:**
```
106         query_fragments = ['CREATE TABLESPACE %s' % pg_quote_identifier(tablespace, 'database')]
107         query_fragments.append("LOCATION '%s'" % location)
108         query = ' '.join(query_fragments)
```

Like this:
```
query = 'CREATE TABLESPACE %s LOCATION %s' % (pg_quote_identifier(tablespace, 'database'), location)
```

**9. login_flags() seems is not used there**

**10. Different SQL code style:**
```
 78     query = """
 79     select spcname, pg_tablespace_location(oid) AS location from pg_tablespace
 80        WHERE spcname = %(tablespace)s
 81     """
```

Better to change it to:
```
query = ("SELECT spcname, pg_tablespace_location(oid) AS location "
         "FROM pg_tablespace WHERE spcname = %(tablespace)s")
```
or
```
query = "SELECT spcname, pg_tablespace_location(oid) AS location "\
        "FROM pg_tablespace WHERE spcname = %(tablespace)s"
```

or something else in upper case for SQL keywords

**11. You can use one of the get_tablespace_info() or tablespace_exists() functions instead both**, they've got similar bodies and logic.
get_tablespace_info() is more general and suitable in this module, I think
```
def get_tablespace_info(cursor, tablespace):
    query = """
    select spcname, pg_tablespace_location(oid) AS location from pg_tablespace
       WHERE spcname = %(tablespace)s
    """
    cursor.execute(query, {'tablespace': tablespace})
    return cursor.fetchone()

def tablespace_exists(cursor, tablespace):
    query = """
    select spcname, pg_tablespace_location(oid) from pg_tablespace
       WHERE spcname = %(tablespace)s
    """
    cursor.execute(query, {'tablespace': tablespace})
    return cursor.rowcount == 1
```
That's all )

If you need some help (e.g., to add any functionality, code refactoring, etc) I'll be glad to help you
Thank you!

File: lib/ansible/modules/database/postgresql/postgresql_tablespace.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+        absent implies that the database should be removed if present.
+    default: present
+    choices: [ "present", "absent" ]
+author: "Ansible Core Team"
+extends_documentation_fragment:
+- postgres
+'''
+RETURN = ''' # '''
+
+EXAMPLES = '''
```
Comment (Andersson007) [inline lib/ansible/modules/database/postgresql/postgresql_tablespace.py:44]: I wrote all notes to the comment

**Review by Andersson007 — CHANGES_REQUESTED** [review]
Hello, @antoinell

**1.What about ALTER TABLESPACE ?** Probably better to add this functionality right now.
```
ALTER TABLESPACE name RENAME TO new_name
ALTER TABLESPACE name OWNER TO new_owner
ALTER TABLESPACE name SET ( tablespace_option = value [, ... ] )
ALTER TABLESPACE name RESET ( tablespace_option [, ... ] )
```

**2. author: "Ansible Core Team"**, Are you sure? I see the "new contributor" status of PR.

**3. Better to add more examples**, for every case

**4. Also it makes sence to add tablespace_option and ownership**(https://www.postgresql.org/docs/11/sql-createtablespace.html)
```
CREATE TABLESPACE tablespace_name
    [ OWNER { new_owner | CURRENT_USER | SESSION_USER } ]
    LOCATION 'directory'
    [ WITH ( tablespace_option = value [, ... ] ) ]
```

**5. Probably better to collect info about the current tablespace and return it by the RETURN block.**
So we could see and use this information by passing the '-vvv' ansible-playbook param.

**6. do_with_password() seems is not used there**

**7. Rename variable to tablespace_info in the tablespace_create() as you did in the next function**
```
112         db_info = get_tablespace_info(cursor, tablespace)
```

**8. It should be changed:**
```
106         query_fragments = ['CREATE TABLESPACE %s' % pg_quote_identifier(tablespace, 'database')]
107         query_fragments.append("LOCATION '%s'" % location)
108         query = ' '.join(query_fragments)
```

Like this:
```
query = 'CREATE TABLESPACE %s LOCATION %s' % (pg_quote_identifier(tablespace, 'database'), location)
```

**9. login_flags() seems is not used there**

**10. Different SQL code style:**
```
 78     query = """
 79     select spcname, pg_tablespace_location(oid) AS location from pg_tablespace
 80        WHERE spcname = %(tablespace)s
 81     """
```

Better to change it to:
```
query = ("SELECT spcname, pg_tablespace_location(oid) AS location "
         "FROM pg_tablespace WHERE spcname = %(tablespace)s")
```
or
```
query = "SELECT spcname, pg_tablespace_location(oid) AS location "\
        "FROM pg_tablespace WHERE spcname = %(tablespace)s"
```

or something else in upper case for SQL keywords

**11. You can use one of the get_tablespace_info() or tablespace_exists() functions instead both**, they've got similar bodies and logic.
get_tablespace_info() is more general and suitable in this module, I think
```
def get_tablespace_info(cursor, tablespace):
    query = """
    select spcname, pg_tablespace_location(oid) AS location from pg_tablespace
       WHERE spcname = %(tablespace)s
    """
    cursor.execute(query, {'tablespace': tablespace})
    return cursor.fetchone()

def tablespace_exists(cursor, tablespace):
    query = """
    select spcname, pg_tablespace_location(oid) from pg_tablespace
       WHERE spcname = %(tablespace)s
    """
    cursor.execute(query, {'tablespace': tablespace})
    return cursor.rowcount == 1
```
**12. It's a good idea to use try-except operators** to catch and handle any exceptions during execution 
```
def tablespace_delete(cursor, tablespace):
    if tablespace_exists(cursor, tablespace):
        query = "DROP TABLESPACE %s" % pg_quote_identifier(tablespace, 'database')
        cursor.execute(query)
        return True
    else:
        return False
```
it may be something like this (note, I added the 'module' argument here):
```
def tablespace_delete(cursor, tablespace, module):
    if tablespace_exists(cursor, tablespace):
        query = "DROP TABLESPACE %s" % pg_quote_identifier(tablespace, 'database')
        try:
            cursor.execute(query)
            return True
       except psycopg2.ProgrammingError as e:
            module.fail_json(msg="Failed to drop tablespace %s: %s" % (tablespace, e))
    else:
        return False
```


That's all )

If you need some help (e.g., to add any functionality, code refactoring, etc) I'll be glad to help you
Thank you!

**Comment by antoinell** [comment]
@Andersson007 I do not have a lot of time available to spend on this pull request and today was the first time I had the ability to get started. I do not mind if someone else beats me to do all the improvements that you suggested in your comment.

**Comment by Andersson007** [comment]
@antoinell , I hope you (or @Dorn-) done it because the module sounds good.

**Comment by Andersson007** [comment]
@antoinell would you mind if anyone completely rewrite this module (of course, you and @Dorn- will be said as co-authors) ?
I want to be the third who want to try luck with it ;)

**Comment by antoinell** [comment]
Hello Andrey,

I would not mind at all as I do not manage to allocate enough time to
finish it.

Antoine

**Comment by Andersson007** [comment]
@antoinell, I'll add you as '- Antoine Levy-Lambert (@antoinell)', ok?

@Dorn- , would you mind if I add you as an co-author in the module description / documentation ? If yes, what name should I use (maybe your real name)?

**Comment by Dorn-** [comment]
@Andersson007 No you could add me as 'Flavien Chantelot'
As @antoinell , I don't have enough time atm to rewrite totally this module :)

**Comment by Andersson007** [comment]
;) Good, I hope to compete it before the next week. I'll inform you both about the PR directly.

**Comment by Andersson007** [comment]
duplicate of https://github.com/ansible/ansible/pull/52914

@gundalow , close this PR, please

**Pull request closed without merging**